content/blog/2023-11-17-argo-configmanagement-plugins.md

```diff
@@ -0,0 +1,142 @@
+in manifests. This allows users to avoid storing secrets directly in git repositories, which is one of the key needs
+of an operational GitOps strategy.
+
+## Implementation
+
+The implementation in the Validated Patterns framework is meant to conform to the mechanism described [here](https://argo-cd.readthedocs.io/en/stable/operator-manual/config-management-plugins/) upstream. Note that the plugin configuration
+actually must live inside the container - either "baked in" to the sidecar image, or injected via configmap. The
+framework supports both options.
+
+Previously, the Validated Patterns `clusterGroup` chart would create three plugins using the CMP 1.0 framework. Only one
+of these was ever used (to the best of our knowledge), `helm-with-kustomize` in the [Industrial Edge](https://github.com/validatedpatterns/industrial-edge) pattern. As of this publication, the kustomize integration is no longer necessary -
+the code was refactored to operate directly on helm value files instead of creating dynamic kustomize patches.
+
+In the `clusterGroup` chart (which is the heart of the Validated Patterns framework), there is a new key, `argoCD`,
+that can optionally be used to implement an arbitrary number of CMP 2.0-style plugins.
+
+For example:
+
+```yaml
+  argoCD:
+    initContainers: []
+    configManagementPlugins:
+      - name: helm-with-kustomize
+        image: quay.io/hybridcloudpatterns/utility-container:latest
+        imagePullPolicy: Always
+        pluginArgs:
+          - '--loglevel=debug'
+        pluginConfig: |
+          apiVersion: argoproj.io/v1alpha1
+          kind: ConfigManagementPlugin
+          metadata:
+            name: helm-with-kustomize
+          spec:
+            preserveFileMode: true
+            init:
+              command: ["/bin/sh", "-c"]
+              args: ["helm dependency build"]
+            generate:
+              command: ["/bin/bash", "-c"]
+              args: ["helm template . --name-template ${ARGOCD_APP_NAME:0:52}
+                -f $(git rev-parse --show-toplevel)/values-global.yaml
+                -f $(git rev-parse --show-toplevel)/values-{{ .Values.clusterGroup.name }}.yaml
+                --set global.repoURL=$ARGOCD_APP_SOURCE_REPO_URL
+                --set global.targetRevision=$ARGOCD_APP_SOURCE_TARGET_REVISION
+                --set global.namespace=$ARGOCD_APP_NAMESPACE
+                --set global.pattern={{ .Values.global.pattern }}
+                --set global.clusterDomain={{ .Values.global.clusterDomain }}
+                --set global.hubClusterDomain={{ .Values.global.hubClusterDomain }}
+                --set global.localClusterDomain={{ coalesce .Values.global.localClusterDomain .Values.global.hubClusterDomain }}
+                --set clusterGroup.name={{ .Values.clusterGroup.name }}
+                --post-renderer ./kustomize"]
+```
+
+`initContainers` is an array of initContainers that will be added to the repo-server pod. In most cases you do not need
```

Comment:
Any correlation between this and the `spec.init` entry?
Do we have any ideas when `initContainers` would make sense to use?

Reply:
Not the argocd `spec.init`. I don't have a strong sense for where this would be needed. The default behavior seems to be pretty reasonable and the main thing you need to happen (copying the existing argocd into the plugin container) already happens by default.